### PR TITLE
Jgfouca/fix fpe testing

### PR DESCRIPTION
**Pull request by jgfouca**
Change list:
1) Fixes SCREAM_FPE so that FPEs are actually on when it is defined (merged this from a previous branch by @bartgol )
2) Adds a test to scripts-tests to ensure that FPEs are thrown when SCREAM_FPE is on
3) Fix all FPEs in P3, none of these was hard or concerning
4) Fix all FPEs in SHOC, there was a floating point overflow that appears to be a real issue. I will annotate it inline.
5) Fix all FPEs in rrtmgp. One of these was very concerning and required a change to the submodule in addition to disabling some tests that started breaking when I put this fix in.
6) Add new SCREAM_AUTOTESTER cmake setting. Allows behavior changes when a run is an AT run such as disabling certain tests.
7) Disable shoc_cld_p3_rrtmgp

### CODE REVIEW

File: components/eam/src/physics/cam/micro_p3.F90
Comment (jgfouca) [inline components/eam/src/physics/cam/micro_p3.F90:2939]: I think these should probably be != 0.

File: components/scream/scripts/scripts-tests
Comment (jgfouca) [inline components/scream/scripts/scripts-tests:356]: Don't need these newlines.

File: components/scream/src/physics/p3/p3_ni_conservation_impl.hpp
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_ni_conservation_impl.hpp:20]: != 0

File: components/scream/src/physics/p3/p3_nr_conservation_impl.hpp
Comment (jgfouca) [inline components/scream/src/physics/p3/p3_nr_conservation_impl.hpp:20]: != 0

File: components/scream/src/physics/share/physics_test_data.hpp
Comment (jgfouca) [inline components/scream/src/physics/share/physics_test_data.hpp:43]: We just want to initialize the Data struct with garbage, it's going to be assigned in the next line anyway. 1 garbage is less likely to be a problem since some of the Data constructors have inv_* members that divide.

File: components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp:273]: Here is the overflow problem on the C++ side. With very small std_s1, the division below can overflow max float.

File: components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp:307]: Another overflow check here.

**Review by jgfouca — COMMENTED** [review]
Annotations.

File: components/scream/src/physics/rrtmgp/tests/rrtmgp_tests.cpp
Comment (jgfouca) [inline components/scream/src/physics/rrtmgp/tests/rrtmgp_tests.cpp:116]: @brhillman , had to disable these once the rrtmgp FPE was fixed.

File: components/scream/tests/uncoupled/rrtmgp/rrtmgp_stand_alone.cpp
Comment (jgfouca) [inline components/scream/tests/uncoupled/rrtmgp/rrtmgp_stand_alone.cpp:281]: @brhillman , had to disable these once the rrtmgp FPE was fixed too.

**Review by jgfouca — COMMENTED** [review]
Some additional notes for Ben.

File: components/eam/src/physics/cam/micro_p3.F90
Comment (ambrad) [reply components/eam/src/physics/cam/micro_p3.F90:2939]: Does the original code actually cause an FPE? I think source_nr must be >= 0. Thus, if sink_nr > source_nr, then sink_nr > 0, causing no issues in source_nr/sink_nr.

File: components/scream/src/physics/p3/p3_main_impl.hpp
Comment (ambrad) [inline components/scream/src/physics/p3/p3_main_impl.hpp:802]: This code would be improved by one overall `if (qc_gt_small.any())`. That would then naturally account for the /0 check.

File: components/eam/src/physics/cam/micro_p3.F90
Comment (jgfouca) [reply components/eam/src/physics/cam/micro_p3.F90:2939]: @ambrad , yes. My methodology was:
1) Build SCREAM with SCREAM_FPE
2) RUN ctest
3) Pick one of the failing tests
4) Run it through gdb, find FPE line, fix it, run it again, repeat until test passes

So, I don't think I touched anything in this PR unless it was to fix an FPE. I can double-check this particular change.

File: components/scream/src/physics/p3/p3_main_impl.hpp
Comment (ambrad) [inline components/scream/src/physics/p3/p3_main_impl.hpp:801]: There are two chunks of code here that depend on two different masks. Setting up two `if (mask.any())` blocks would be natural and increase performance. As a side effect, it would take care of the /0 case.

File: components/scream/src/physics/p3/p3_main_impl.hpp
Comment (ambrad) [inline components/scream/src/physics/p3/p3_main_impl.hpp:828]: Same here.

File: components/scream/src/physics/p3/p3_ni_conservation_impl.hpp
Comment (ambrad) [reply components/scream/src/physics/p3/p3_ni_conservation_impl.hpp:20]: See my comment above re: the same code in the F90. Sorry if I'm just confused.

File: components/eam/src/physics/cam/micro_p3.F90
Comment (jgfouca) [reply components/eam/src/physics/cam/micro_p3.F90]: @ambrad , I stand corrected. This check was not necessary.

File: components/scream/src/physics/p3/p3_nr_conservation_impl.hpp
Comment (ambrad) [reply components/scream/src/physics/p3/p3_nr_conservation_impl.hpp:20]: Same as sink_ni.

File: components/scream/src/physics/p3/p3_ni_conservation_impl.hpp
Comment (jgfouca) [reply components/scream/src/physics/p3/p3_ni_conservation_impl.hpp:20]: @ambrad , this change does appear to be necessary. When I remove it, I get FPEs in all the p3_tests.

File: components/scream/src/physics/p3/p3_nr_conservation_impl.hpp
Comment (jgfouca) [reply components/scream/src/physics/p3/p3_nr_conservation_impl.hpp:20]: @ambrad , this change does appear to be necessary. When I remove it, I get FPEs in all the p3_tests.

File: components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp
Comment (ambrad) [inline components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp:161]: Looks like there is the same opportunity here to do one big `condition.any()` check to either run or not run a big chunk of code. That will then naturally solve the /0 problem.

File: components/eam/src/physics/cam/micro_p3.F90
Comment (jgfouca) [reply components/eam/src/physics/cam/micro_p3.F90:2939]: @ambrad sorry for the confusion; this change does appear to be necessary (I tested my non-FPE build on accident when I said it was not necessary).

File: components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp
Comment (ambrad) [reply components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp:273]: If possible, use a numeric_limits<Real> function so this threshold works for single precision too.

File: components/eam/src/physics/cam/micro_p3.F90
Comment (ambrad) [reply components/eam/src/physics/cam/micro_p3.F90:2939]: Is source_nr going negative? If it is, we should bring this to the attention of Aaron. Otherwise, I'm not seeing how this code triggers an FPE.

Comment (jgfouca) [reply components/eam/src/physics/cam/micro_p3.F90:2939]: @ambrad , I remove the `.and. sink_nr.gt.0._rtype` and added a print inside the if block. The values printed immediately before the FPE crash were:

```
sink_nr= -0.00000000000000000000000000000000E+00 source_nr= -3.63189856026829738233629996102536E-01
```

**Comment by ambrad** [comment]
I believe dz_zi(k) is a fatal error, so the dz_zi cases might just be an artifact of the testing inputs. If so, it's definitely fair to make the dz testing inputs always > 0.

File: components/eam/src/physics/cam/micro_p3.F90
Comment (ambrad) [reply components/eam/src/physics/cam/micro_p3.F90:2939]: @AaronDonahue is source_nr < 0 and sink_nr = 0 a physically valid state? If not, we need to figure out what's leading to it.

File: components/eam/src/physics/cam/shoc.F90
Comment (bartgol) [reply components/eam/src/physics/cam/shoc.F90:2858]: You could use `tiny(X)` (with X any number of kind rtype). It seems to be part of the fortran standard, at least f95 and later.

File: components/scream/src/physics/p3/p3_ni_conservation_impl.hpp
Comment (bartgol) [reply components/scream/src/physics/p3/p3_ni_conservation_impl.hpp:20]: I think, as stated by Andrew in another comment, that we might have to check with Aaron. A negative sink is a source, but since we have two distinct vars for sink and sources, I would expect both of them to be positive. If we are not "wasting" time to clip them to 0, then I understand, and this check is fine, but if they were not supposed to get negative in the first place, then we should do a `!=0` check (so that a <0 val can still generate errors, and reveal bugs).

File: components/scream/src/physics/p3/tests/p3_rain_sed_unit_tests.cpp
Comment (bartgol) [inline components/scream/src/physics/p3/tests/p3_rain_sed_unit_tests.cpp:55]: Maybe there was a reason to test 0.0? Like stress-testing some conditionals?

File: components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp
Comment (bartgol) [inline components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp:154]: This expression is quite complicated. Is it guaranteed that if condition is true, then the denominator is nonzero? To me, it's hard to tell. Might be that with this mod the test passes, but I can't tell if it just "happens" to pass, or if it will "always" pass, even when we add more scream tests (e.g., fully coupled scream, with real data).

File: components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp
Comment (bartgol) [inline components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp:301]: You could do `if (not equal.none())`. Mighr help save a register on GPU.

**Comment by jgfouca** [comment]
Hi all,

I'm going to need some help getting this across the finish line. Unhiding our FPEs has revealed some concerns/errors in the code that are real. Here are the points on which I'd like some help/feedback.
1) @AaronDonahue , in P3 for nr_conservation and ni_conservation, we are sometimes seeing the source=0 and the sink<0. Is this valid/expected? This appears to only happen for unit tests, not p3_run_and_cmp, so maybe the unit test inputs are bad?
2) @bogensch , in SHOC for shoc_assumed_pdf_compute, the value of std_s is getting small enough that a subsequent division by std_s is causing a result that overflows a double. Through printing and trial&error, I found that limiting std_s to be no smaller than `2.0E-155` (just treat it like zero if it is) fixes this problem. I don't know if this is how we want to handle this situation. This appears to only happen for unit tests, not shoc_run_and_cmp, so maybe the unit test inputs are bad?
3) @bartgol , when I initially developed this PR, I did not have `-DSCREAM_DYNAMICS_DYCORE=HOMME` , so I did not notice until the AT returned that there are some FPEs coming from Homme as well. I had to do a fair bit of digging but I found out some info that is hopefully useful:
The failing stack trace is as follows:
```
test
  DynamicsDrivenGridsManager::build_grids
    homme_grid_mod::init_grids_f90
      dyn_grid_mod::dyn_grid_init
        prim_driver_base::prim_init1_geometry
          geometry_mod::set_area_correction_map0
            cube_mod::cube_init_atomic
              cube_mod::metric_atomic (FPE on this line: elem%Dinv = elem%Dinv / sqrt(alpha) )
```
The FPE occurs because `alpha` is zero. `alpha` is zero because in geometry_mod::set_area_correction_map0, the value that becomes `alpha` lower in stack is computed via `area(1) = domain_size/area(1)  ! ratio correction`. The value is zero because `domain_size` is zero. `domain_size` comes from the `physical_constants` module and it should have been initialized by `namelist_mod::readnl` but apparently this subroutine was never called. Based on this, it seems like the homme_pd_remap_tests need to call `namelist_mod::readnl`. If this is correct, what's the easiest way to do this? I see a couple nl files living alongside this test `theta-ndays.nl  theta-nmax.nl`.

**Comment by AaronDonahue** [comment]
Per:

```
@AaronDonahue , in P3 for nr_conservation and ni_conservation, we are sometimes seeing the source=0 and the sink<0. Is this valid/expected? This appears to only happen for unit tests, not p3_run_and_cmp, so maybe the unit test inputs are bad?
```

Looking at the formulation of both source and sink in nr_conservation and ni_conservation, it should not be possible to get sink<0.  In both cases the sink term is a sum of what should be positive tendencies.  You mentioned that you see negative sink in the unit tests?  I'll start there and investigate what is going on.  I'll comment back here if/when I find something.

**Comment by AaronDonahue** [comment]
@jgfouca , I think I found the source of the issue with negative source/sinks.  One of the test cases in p3_tests was not setting `dt` in the FortranDataIterator, which was defaulting to `dt=-1`.  The fix looks like:

```
diff --git a/components/scream/src/physics/p3/p3_f90.cpp b/components/scream/src/physics/p3/p3_f90.cpp
index 7cbeb8541b..8a1277c846 100644
--- a/components/scream/src/physics/p3/p3_f90.cpp
+++ b/components/scream/src/physics/p3/p3_f90.cpp
@@ -167,6 +167,7 @@ int test_p3_init () {
 
 int test_p3_ic (bool use_fortran) {
   const auto d = ic::Factory::create(ic::Factory::mixed);
+  d->dt = 300.0;
   p3_init();
   p3_main(*d, use_fortran);
   P3GlobalForFortran::deinit();
```

**Comment by AaronDonahue** [comment]
I have also found another location where we can get negative sink values, see https://github.com/E3SM-Project/scream/blob/master/components/scream/src/physics/p3/p3_rain_self_collection_impl.hpp#L35

because we use 2 - exp(X) it is possible for X to be large enough to cause this to go negative.  This bug is also present in the F90 code, as expected.  I think this is something worth investigating.

**Comment by AaronDonahue** [comment]
Depdending on the results of Issue #1094 , this is a possible fix that removes the remaining sink/source<0 issues, according to my tests:

```
diff --git a/components/scream/src/physics/p3/p3_rain_self_collection_impl.hpp b/components/scream/src/physics/p3/p3_rain_self_collection_impl.hpp
index 361c87a1b3..3520e8db32 100644
--- a/components/scream/src/physics/p3/p3_rain_self_collection_impl.hpp
+++ b/components/scream/src/physics/p3/p3_rain_self_collection_impl.hpp
@@ -34,8 +34,16 @@ void Functions<S,D>
     if (dum2_gt_dum1.any()) {
       dum.set(dum2_gt_dum1, 2 - exp(2300 * (dum2-dum1)));
     }
+    const auto dum_neg = dum < 0 && qr_incld_not_small;
+    if (dum_neg.any()) {
+      dum.set(dum_neg,0.0);
+    }
```

and

```
diff --git a/components/eam/src/physics/cam/micro_p3.F90 b/components/eam/src/physics/cam/micro_p3.F90
index 2468b643d4..45ddb02360 100644
--- a/components/eam/src/physics/cam/micro_p3.F90
+++ b/components/eam/src/physics/cam/micro_p3.F90
@@ -2719,6 +2719,8 @@ subroutine rain_self_collection(rho,qr_incld,nr_incld,    &
       else if (dum2.ge.dum1) then
          dum = 2._rtype - bfb_exp(2300._rtype*(dum2-dum1))
       endif
+      ! Clip dum so that it can't be <0
+      dum = max(dum,0.0);
```

**Comment by jgfouca** [comment]
@AaronDonahue , your fix worked great! I've removed the extra checks I added to ni|nr_conservation. I also added a check to p3|shoc_main to ensure that parameters that are not automatically set by the p3|shoc:::FortranData Factories are set to something reasonable by the time p3|shoc_main is called.

**Comment by jgfouca** [comment]
@bartgol , my latest push takes care of the homme FPE. Please review.

File: components/scream/src/physics/rrtmgp/tests/rrtmgp_tests.cpp
Comment (brhillman) [reply components/scream/src/physics/rrtmgp/tests/rrtmgp_tests.cpp:116]: This is a little worrisome. We should be able to keep these lines and just update the baselines, but let me check with the RRTMGP offline tests and make sure the changes still give the right answers to machine precision before we do that.

**Review by bartgol — APPROVED** [review]

**Review by brhillman — COMMENTED** [review]
I think the RRTMGP fixes look good now. I put the changes in a PR upstream to our RRTMGP repo, assuming everything checks out I think we should switch over to main once I get that PR integrated.

File: components/scream/src/physics/p3/tests/p3_rain_sed_unit_tests.cpp
Comment (jgfouca) [reply components/scream/src/physics/p3/tests/p3_rain_sed_unit_tests.cpp:55]: @bartgol , yes, I'm sure that was the intention here. The "real" calls to compute_rain_fall_velocity occur within a block of code in rain_sedimentation that's already protected by qr_small checks. I did not want to add a check for the sake of unit testing that would be redundant and potentially reduce performance in the context of the full system.

File: components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp
Comment (jgfouca) [reply components/scream/src/physics/shoc/shoc_assumed_pdf_impl.hpp:154]: Good question. I think if `w1_2 - w1_1 > w_thresh` then `tmp_val_1 - tmp_val_2` should never be zero.

File: components/eam/src/physics/cam/shoc.F90
Comment (PeterCaldwell) [reply components/eam/src/physics/cam/shoc.F90:2858]: Hmm. One question is why std_s would ever approach 0. s is the dry static energy, so is constructed from a sum of temperature and vapor variances... I'm not sure why these would ever be zero. So: are we sure we're getting overflow in realistic cases, or just with random data for unit tests? 

Beyond the question of whether the inputs causing the problem are reasonable, I think your hack is fine as a temporary solution. A better solution might be to notice that the problem calculation always appears as R=s/std_s. The right thing to do would be to use L'Hospital's rule to decide whether s or std_s dominates in this case. I think s should be physically bounded by the temperature and water vapor values seen on planet Earth, while std_s could get arbitrarily large (and is a standard deviation so is always positive), so my bet on L'Hospital is that R will go to infinity as std_s goes to zero. 

Note as well that R is only used in erf(R) and exp(-R) types of expressions, which quickly converge to 1 or 0 as R gets big. This means that we can replace the overflowing R with a big number without impacting the solution. So maybe replace `if (std_s .gt. 2.0E-155_rtype)` with `std_s = min(std_s, tiny(rtype) )`? I think this would give you the C=1 and qn=s  that you're prescribing in the `else` for the `std_s > 2e-155` conditional. So you get rid of the overflow *and* the conditional!

File: components/scream/src/physics/shoc/shoc_calc_shoc_varorcovar_impl.hpp
Comment (jgfouca) [reply components/scream/src/physics/shoc/shoc_calc_shoc_varorcovar_impl.hpp]: @ambrad , you're right. The issue here was not so much that dz_zi was zero, it is that we were not protecting these blocks of code with `any()` checks on the range masks.

File: components/eam/src/physics/cam/shoc.F90
Comment (jgfouca) [reply components/eam/src/physics/cam/shoc.F90:2858]: @PeterCaldwell , note that `tiny(rtype)` would be somewhere in the range of 1e-308 for doubles, so much much smaller than 2e-155. Since it is the square of `s/std_s` that is overflowing, we'd need to check against something closer to `sqrt(tiny(rtype))`, depending on size of s.

Comment (PeterCaldwell) [reply components/eam/src/physics/cam/shoc.F90:2858]: Ok, I suggested using tiny(rtype) rather than 2e-155 because hardcoded numbers look ugly rather than for any good reason. The bigger point I was trying to make is that it shouldn't matter whether you use 2e-155 or 1e-30 or tiny(rtype) because all of them will get mapped to approximately identical values of erf(R) or exp(-R). So feel free to replace my ill-advised tiny statements with 2e-155 or whatever value you like.

Comment (bartgol) [reply components/eam/src/physics/cam/shoc.F90:2858]: On a different note, shoc.F90 is not going to be supported in scream for much longer, so maybe whatever hack works is fine?

Comment (bartgol) [reply components/eam/src/physics/cam/shoc.F90:2858]: But given that you need cxx-f90 bfb-ness, and we want to avoid errors in cxx even when the f90 goes away (if ever), then I encourage a formulation based on a not-hard-coded number. Jim, I think something like `bfb_sqrt(tiny(rtype))` as you suggested is fine.

Comment (jgfouca) [reply components/eam/src/physics/cam/shoc.F90]: @PeterCaldwell , I was able to come up with a less-hacky min value, `bfb_sqrt(tiny(1._rtype)) * 100`. Trying to get rid of the conditional entirely led to some subtle non-bfbness between f90 and cxx that I'm not sure I want to track down. The cxx version of shoc_assumed_pdf stuff has diverged significantly from fortran.

Comment (PeterCaldwell) [reply components/eam/src/physics/cam/shoc.F90]: ok, that's fine with me.

**Comment by jgfouca** [comment]
It looks like new FPEs have made it onto master (undetected) which is caused AT to fail my PR (it does an upstream merge of master, bringing in the new FPE problems which get now get detected on my PR). Consulting with @brhillman since the FPE is in the new test `shoc_cld_p3_rrtmgp_ut_np1_omp1`.

**Comment by brhillman** [comment]
The current issue (after merging in master) with the shoc_p3_cld_rrtmgp test is that mu0/cos_zenith has columns that are zero-valued. This is totally reasonable, mu0 <= 0 just means the sun is at or below the horizon, but the solar radiative transfer code doesn’t know what to do with this. Currently, RRTMGP assumes only daytime columns are passed to the code so that a bunch of conditionals are avoided within the kernels. In order to deal with this, and to save some computations, the strategy in EAM has always been to subset the inputs to daytime-only columns before passing to the solar radiative transfer codes, and then to expand the fluxes back to all columns afterwards. The alternative is to add conditionals to simply pass over the columns with mu0 <= 0, which would avoid data copies and make the higher level code cleaner but would mean more computations relative to the subsetting approach, and might mean fairly intrusive modifications to the RRTMGP code (not a huge deal). Regardless, I think both of these strategies are beyond of the scope of this PR, so we should disable this test from the autotester, merge the PR, and just let the test register a fail on the dashboard until we fix it (which should be short order, but I don’t think we want to make these side effect changes in Jim’s PR, nor hold up the PR any more).

**Comment by jgfouca** [comment]
Thx, @brhillman

**Comment by jgfouca** [comment]
At long last this is ready to merge.

**Pull request merged**
